src/test/fuzz/versionbits.cpp

```diff
@@ -0,0 +1,345 @@
+// Copyright (c) 2020-2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <chain.h>
+#include <chainparams.h>
+#include <consensus/params.h>
+#include <primitives/block.h>
+#include <versionbits.h>
+
+#include <test/fuzz/FuzzedDataProvider.h>
+#include <test/fuzz/fuzz.h>
+#include <test/fuzz/util.h>
+
+#include <cstdint>
+#include <limits>
+#include <memory>
+#include <vector>
+
+namespace {
+class TestConditionChecker : public AbstractThresholdConditionChecker
+{
+private:
+    mutable ThresholdConditionCache m_cache;
+    const Consensus::Params dummy_params{};
+
+public:
+    const int64_t m_begin = 0;
+    const int64_t m_end = 0;
+    const int m_period = 0;
+    const int m_threshold = 0;
+    const int m_bit = 0;
+
+    TestConditionChecker(int64_t begin, int64_t end, int period, int threshold, int bit)
+        : m_begin{begin}, m_end{end}, m_period{period}, m_threshold{threshold}, m_bit{bit}
+    {
+        assert(m_period > 0);
+        assert(0 <= m_threshold && m_threshold <= m_period);
+        assert(0 <= m_bit && m_bit <= 32 && m_bit < VERSIONBITS_NUM_BITS);
+    }
+
+    bool Condition(const CBlockIndex* pindex, const Consensus::Params& params) const override { return Condition(pindex->nVersion); }
+    int64_t BeginTime(const Consensus::Params& params) const override { return m_begin; }
+    int64_t EndTime(const Consensus::Params& params) const override { return m_end; }
+    int Period(const Consensus::Params& params) const override { return m_period; }
+    int Threshold(const Consensus::Params& params) const override { return m_threshold; }
+
+    ThresholdState GetStateFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateFor(pindexPrev, dummy_params, m_cache); }
+    int GetStateSinceHeightFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateSinceHeightFor(pindexPrev, dummy_params, m_cache); }
+    BIP9Stats GetStateStatisticsFor(const CBlockIndex* pindexPrev) const { return AbstractThresholdConditionChecker::GetStateStatisticsFor(pindexPrev, dummy_params); }
+
+    bool Condition(int64_t version) const
+    {
+        return ((version >> m_bit) & 1) != 0 && (version & VERSIONBITS_TOP_MASK) == VERSIONBITS_TOP_BITS;
```

Comment:
I'm curious why you decided to right shift the version message instead of matching how `VersionBitsConditionChecker` left shifts the bit to compare to the block's version. 

Also why the `version` param here is an `int64_t` when `CBlockHeader.nVersion` is `int32_t`. 

I don't think (?) any of these cause problems, but I think we want the test checker to match the version bits checker as closely as possible?

Reply:
I don't think there's any good reason for either of those changes.